File: Classes/MenuViewController.m
//
//  MenuViewController.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/12/09.
//  Revised by Krishna Narayan on 9/17/09
//		--Updated graphics
//		--Added tell a friend
//		--Changed app name to Kids Learn To Tell Time
//  Modified by SVV Satyanarayana, under contract to NSC Partners LLC on 28/04/10.
//  Copyright 2009-2013 NSC Partners LLC. Copyright 2014 One Step Ahead Apps, LLC. All rights reserved.
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Copyright 2026 Island Innovation LLC.  All rights reserved.
//

#import <QuartzCore/QuartzCore.h>
#import "MenuViewController.h"
#import "ActivityViewController.h"
#import "KidsTimeFunAppState.h"
#import "HelpViewController.h"
#import "TellAFriendViewController.h"
#import "TopScoresActivitySelector.h"
#import "SettingsModalViewController.h"
#import <FloopSDK/FloopSDK.h>

@implementation MenuViewController

@synthesize tellTimeButton;
@synthesize setTimeButton;
@synthesize elapsedTimeButton;
@synthesize tellTimeAfterButton;
@synthesize tellTimeBeforeButton;
@synthesize mixedModeButton;
@synthesize topScoresButton;
@synthesize helpButton;
@synthesize tellAFriendButton;
@synthesize choiceActivityType;
@synthesize clockView;
@synthesize logoImageView;
@synthesize clipArtImageView;
@synthesize clipArtView;

@synthesize topScoresActVC;
@synthesize settingsVC;
@synthesize helpVC;
@synthesize activityVC;

NSTimer *clockTimer;
NSTimer *clipArtTimer;

/*
 // The designated initializer.  Override if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
 - (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
 if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
 // Custom initialization
 }
 return self;
 }
 */

/*
 // Implement loadView to create a view hierarchy programmatically, without using a nib.
 - (void)loadView {
 }
 */

// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad {
    //Put Title - App Name/Menu
    self.title = kStrAppTitle;
    //Customize navItem - add settings button and top cores button
    UIBarButtonItem *settingsBarButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Settings"] style:UIBarButtonItemStyleBordered target:self action:@selector(settingsActivated)];
    self.navigationItem.rightBarButtonItem = settingsBarButton;
    UIBarButtonItem *topScoresBarButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Top Scores"] style:UIBarButtonItemStyleBordered target:self action:@selector(topScoresButtonPressed:)];
    self.navigationItem.leftBarButtonItem = topScoresBarButton;
    
    
    //[settingsBarButton release];
    //Custom Back Button - home button
    UIBarButtonItem *homeBarButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:kImgHome] style:UIBarButtonItemStylePlain target:self action:@selector(goHome)];
    [[UIBarButtonItem appearance] setTintColor:[UIColor colorWithRed:0.055 green:0.478 blue:0.996 alpha:1.000]];
    self.navigationItem.backBarButtonItem = homeBarButton;
    [homeBarButton release];
    //set activity Type to default
    choiceActivityType.selectedSegmentIndex = [KidsTimeFunAppState sharedState].activityType;
    clipArtTimer = nil;
    clipArtImageView.frame = logoImageView.frame = CGRectMake(0, 0, clipArtView.frame.size.width, clipArtView.frame.size.height);
    clipArtImageView.contentMode = logoImageView.contentMode = UIViewContentModeScaleAspectFit;
    [clipArtView addSubview:logoImageView];
    clipArtTimer = [NSTimer scheduledTimerWithTimeInterval:2 target:self selector:@selector(changeClipArt) userInfo:NULL repeats:YES];
    [super viewDidLoad];
}

- (void) viewWillAppear:(BOOL)animated {
    //set labels for activity type choices
    NSString *tmpTitle = [NSString stringWithFormat:kStrVarMaxQuestions, [KidsTimeFunAppState sharedState].maxQuestions];
    [choiceActivityType setTitle:tmpTitle forSegmentAtIndex:kActTypeNumbered];
    int minutes = [KidsTimeFunAppState sharedState].maxTimeInSeconds/60;
    if (minutes == 1) tmpTitle = [NSString stringWithFormat:kStrOneMinute];
    else tmpTitle = [NSString stringWithFormat:kStrVarMaxMinutes, minutes];
    [choiceActivityType setTitle:tmpTitle forSegmentAtIndex:kActTypeTimed];
    //set timer
    [self refreshClock];
    clockTimer = nil;
    clockTimer = [NSTimer scheduledTimerWithTimeInterval:1 target:self selector:@selector(refreshClock) userInfo:NULL repeats:YES];
    [super viewWillAppear:animated];
}

- (void) viewWillDisappear:(BOOL)animated {
    [clockTimer invalidate];
    clockTimer = nil;
    [super viewWillDisappear:animated];
}

/*
 // Override to allow orientations other than the default portrait orientation.
 - (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
 // Return YES for supported orientations
 return (interfaceOrientation == UIInterfaceOrientationPortrait);
 }
 */

- (void)didReceiveMemoryWarning {
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

- (void)viewDidUnload {
    [clipArtTimer invalidate];
    clipArtTimer = nil;
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (void)dealloc {
    [tellTimeButton release];
    [setTimeButton release];
    [elapsedTimeButton release];
    [tellTimeAfterButton release];
    [tellTimeBeforeButton release];
    [mixedModeButton release];
    [topScoresButton release];
    [helpButton release];
    [tellAFriendButton release];
    [choiceActivityType release];
    [clockView release];
    [logoImageView release];
    [clipArtImageView release];
    [clipArtView release];
    [super dealloc];
}


#pragma mark -
#pragma mark Custom Event Handlers

- (void) changeClipArt {
    RandomInteger *randomInt = [[RandomInteger alloc] initWithRange:kClipArtFileRangeLow To:kClipArtFileRangeHigh];
    clipArtImageView.image = [UIImage imageNamed:[NSString stringWithFormat:kClipArtFileMask, randomInt.randomInteger,kClipArtFileType]];
    [clipArtView replaceSubview:[[clipArtView subviews] objectAtIndex:0] withSubview:clipArtImageView transition:kCATransitionPush direction:kCATransitionFromLeft duration:0.10];
    [randomInt release];
}

- (void) refreshClock {
    //set clock
    //create date components for current date
    //NSDateComponents *dateComponents = ;
    //Hours first
    NSCalendar *calendar = [NSCalendar autoupdatingCurrentCalendar];
    NSDateComponents *timeComponents = [calendar components:( NSHourCalendarUnit | NSMinuteCalendarUnit | NSSecondCalendarUnit ) fromDate:[NSDate date]];
    clockView.hours = [timeComponents hour];
    //Then AM/PM
    //clockView.PM = [[NSDate date] AMSymbol];
    //rounded to nearest multiple of timeInterval
    clockView.minutes = [timeComponents minute];
    //No seconds for now
    clockView.seconds = [timeComponents second];
    clockView.showSeconds = YES;
    //other clock properties
    clockView.showClockAsAnalog = YES;
    clockView.showMinutesOffsetInHoursHand = YES;
    clockView.showAMPM = NO;
    clockView.showDayNight = NO;
    [clockView setNeedsDisplay];
}

- (IBAction) tellTimeButtonPressed: (id) sender {
    //ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActTellTime;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //[activityVC release];
}

- (IBAction) setTimeButtonPressed: (id) sender {
    //	ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActSetTime;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //	[activityVC release];
}

- (IBAction) elapsedTimeButtonPressed: (id) sender {
    //	ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActElapsedTime;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //	[activityVC release];
}

- (IBAction) tellTimeAfterButtonPressed: (id) sender {
    //	ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActTimeAfter;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //	[activityVC release];
}

- (IBAction) tellTimeBeforeButtonPressed: (id) sender {
    //	ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActTimeBefore;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //	[activityVC release];
}

- (IBAction) mixedModeButtonPressed: (id) sender {
    //	ActivityViewController *activityVC = [[ActivityViewController alloc] initWithNibName:kNibActivity bundle:nil];
    [KidsTimeFunAppState sharedState].activity = kActMixed;
    //	activityVC.activity = [KidsTimeFunAppState sharedState].activity;
    //	activityVC.activityType = [KidsTimeFunAppState sharedState].activityType;
    //	activityVC.activityLevel = [KidsTimeFunAppState sharedState].activityLevel;
    [self.navigationController pushViewController:activityVC animated:YES];
    //	[activityVC release];
}

- (IBAction) topScoresButtonPressed: (id) sender {
    //	TopScoresActivitySelector *topScoresActVC = [[TopScoresActivitySelector alloc] initWithNibName:@"TopScoresActivitySelectorView" bundle:nil];
    [self.navigationController pushViewController:self.topScoresActVC animated:YES];
    
    //	[topScoresActVC release];
}

- (IBAction) helpButtonPressed: (id) sender {
    //HelpViewController *helpVC = [[HelpViewController alloc] initWithNibName:kNibHelp bundle:nil];
    [self.navigationController pushViewController:self.helpVC animated:YES];
    //[helpVC release];
}

- (IBAction) tellAFriendButtonPressed: (id) sender {
    
    /*
     NSString *recipients = @"mailto:  ?subject=Learn to Tell Time";
     NSString *body = @"&body=Please check out this really cool app:  http://itunes.apple.com/WebObjects/MZStore.woa/wa/viewSoftware?id=318350766&mt=8";
     
     
     NSString *email = [NSString stringWithFormat:@"%@%@", recipients, body];
     email = [email stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
     
     [[UIApplication sharedApplication] openURL:[NSURL URLWithString:email]];
     
     TellAFriendViewController *tellAFriendVC = [[TellAFriendViewController alloc] initWithNibName:kNibTellAFriend bundle:nil];
     [self.navigationController pushViewController:tellAFriendVC animated:YES];
     [tellAFriendVC release];
     */
    Class mailClass = (NSClassFromString(@"MFMailComposeViewController"));
    if (mailClass != nil)
    {
        // We must always check whether the current device is configured for sending emails
        if ([mailClass canSendMail])
        {
            [self displayComposerSheet];
        }
        else
        {
            [self launchMailAppOnDevice];
        }
    }
    else
    {
        [self launchMailAppOnDevice];
    }
}


- (IBAction) setActivityType: (id) sender {
    [KidsTimeFunAppState sharedState].activityType = (int) [sender selectedSegmentIndex];
}

- (IBAction) settingsActivated {
    if ([self.navigationController topViewController] == self) {
        [self.navigationController pushViewController:settingsVC animated:YES];
    }
}

//#pragma mark -
//#pragma mark DissmissSettings Protocol Methods
//
//- (void) didDismissSettings: (id)sender {
//	[self dismissModalViewControllerAnimated:YES];
//}

#pragma mark -
#pragma mark Compose Mail

// Displays an email composition interface inside the application. Populates all the Mail fields.
-(void)displayComposerSheet
{
    MFMailComposeViewController *picker = [[MFMailComposeViewController alloc] init];
    picker.mailComposeDelegate = self;
    
    [picker setSubject:@"Kids Time Fun App"];
    
    
    // Set up recipients
    //NSArray *toRecipients = [NSArray arrayWithObject:@"first@example.com"];
    
    //[picker setToRecipients:toRecipients];
    
    
    // Fill out the email body text
    NSString *emailBody = @"Please try this really cool app:  http://itunes.apple.com/WebObjects/MZStore.woa/wa/viewSoftware?id=318350766";
    [picker setMessageBody:emailBody isHTML:NO];
    
    [self presentViewController:picker animated:YES completion:nil];
    [picker release];
}


// Dismisses the email composition interface when users tap Cancel or Send. Proceeds to update the message field with the result of the operation.
- (void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error
{
    // Notifies users about errors associated with the interface
    switch (result)
    {
        case MFMailComposeResultCancelled:
            message = @"User cancelled";
            break;
        case MFMailComposeResultSaved:
            message = @"Your information saved successfully";
            break;
        case MFMailComposeResultSent:
            message = @"Your friends were informed about this application";
            break;
        case MFMailComposeResultFailed:
            message = @"Sorry, I couldn't inform your friend. Try again";
            break;
        default:
            message = @"Result: not sent";
            break;
    }
    [self dismissViewControllerAnimated:YES completion:nil];
    UIAlertView* av = [[UIAlertView alloc] initWithTitle:@"Tell A Friend" message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [av show];
    [av release];
}


#pragma mark -
#pragma mark Workaround

// Launches the Mail application on the device.
-(void)launchMailAppOnDevice
{
    NSString *recipients = @"mailto:?subject=Learn To Tell Time--Kids iPhone/iPod/iPad App!";
    NSString *body = @"&body=Please try this really cool app:  http://itunes.apple.com/WebObjects/MZStore.woa/wa/viewSoftware?id=318350766";
    
    NSString *email = [NSString stringWithFormat:@"%@%@", recipients, body];
    email = [email stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:email]];
}

// Application Launcher
- (IBAction) launchApp:(id)sender {
    
    UIButton *buttonClicked = (UIButton *)sender;
    
    NSString* appLauchFilePath = [[NSBundle mainBundle] pathForResource:@"AppLaunchInfo" ofType:@"plist"];
    NSURL *appLaunchFileURL = [NSURL fileURLWithPath:appLauchFilePath];
    NSArray *appLauchInfo = [NSArray arrayWithContentsOfURL:appLaunchFileURL];
    
    UIApplication *ourApplication = [UIApplication sharedApplication];
    NSInteger ourAppIndex = buttonClicked.tag-700;
    NSDictionary *ourApp = appLauchInfo[ourAppIndex];
    NSURL *AppLaunchURL = [NSURL URLWithString:ourApp[@"AppLaunchURL"]];
    NSURL *appStoreURL = [NSURL URLWithString:ourApp[@"AppStoreURL"]];
    
    if ([ourApplication canOpenURL:AppLaunchURL]) {
        [ourApplication openURL:AppLaunchURL];
    }
    else {
        [[FloopSdkManager sharedInstance] showParentalGate:^(BOOL success) {
            if(success)
            {
                [ourApplication openURL:appStoreURL];
            }
            ;}];
    }
}
@end

File: Classes/MenuViewController.swift
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Revised by Krishna Narayan on 6/3/26 — Using Claude changed to 1st, 2nd, and 3rd grade levels, belts are earned not selected, added adaptive weak-drilling algorithm to rectify mistakes and build proficiency after initially providing randomized problems for activities
// Copyright 2026 Island Innovation LLC.  All rights reserved.

import UIKit
import QuartzCore

@objc(MenuViewController)
class MenuViewController: UIViewController {

    @IBOutlet var tellTimeButton: UIButton?
    @IBOutlet var setTimeButton: UIButton?
    @IBOutlet var elapsedTimeButton: UIButton?
    @IBOutlet var mixedModeButton: UIButton?
    @IBOutlet var tellTimeAfterButton: UIButton?
    @IBOutlet var tellTimeBeforeButton: UIButton?
    @IBOutlet var topScoresButton: UIButton?
    @IBOutlet var choiceActivityType: UISegmentedControl!
    @IBOutlet var clockView: ClockView!
    @IBOutlet var logoImageView: UIImageView?
    @IBOutlet var clipArtImageView: UIImageView?
    @IBOutlet var clipArtView: TransitionView?
    @IBOutlet var topScoresActVC: TopScoresActivitySelector!
    @IBOutlet var settingsVC: SettingsModalViewController!
    @IBOutlet var activityVC: ActivityViewController!

    private var clockTimer: Timer?
    private var clipArtTimer: Timer?
    private let playerChip = UIButton(type: .system)
    private static var didPresentPickerThisLaunch = false

    override func viewDidLoad() {
        super.viewDidLoad()
        installLegacyScaling(topAligned: true)
        title = kStrAppTitle
        edgesForExtendedLayout = []

        // Tappable player chip (avatar + name) as the title — taps to switch student.
        playerChip.addTarget(self, action: #selector(playerChipTapped), for: .touchUpInside)
        navigationItem.titleView = playerChip
        updatePlayerChip()

        let settingsBtn = UIBarButtonItem(image: UIImage(systemName: "gearshape"), style: .plain, target: self, action: #selector(settingsActivated))
        settingsBtn.accessibilityLabel = "Settings"
        navigationItem.rightBarButtonItem = settingsBtn
        let topScoresBtn = UIBarButtonItem(image: UIImage(systemName: "trophy"), style: .plain, target: self, action: #selector(topScoresButtonPressed(_:)))
        topScoresBtn.accessibilityLabel = kStrRankBelts
        navigationItem.leftBarButtonItem = topScoresBtn

        UIBarButtonItem.appearance().tintColor = UIColor(red: 0.055, green: 0.478, blue: 0.996, alpha: 1)
        let backBtn = UIBarButtonItem(image: UIImage(systemName: "house"), style: .plain, target: nil, action: nil)
        backBtn.accessibilityLabel = "Home"
        navigationItem.backBarButtonItem = backBtn

        // The Questions/Minutes selector is gone — round count and timing are now
        // decided by the belt-progression engine, not chosen on the main screen.
        choiceActivityType.isHidden = true

        // VoiceOver: give each activity launcher a clear spoken name, and hide the
        // purely decorative logo / rotating clip art from the rotor.
        let buttonLabels: [(UIButton?, String)] = [
            (tellTimeButton, "Tell Time"), (setTimeButton, "Set the Time"),
            (elapsedTimeButton, "Elapsed Time"), (mixedModeButton, "Mixed Practice"),
            (tellTimeAfterButton, "Time After"), (tellTimeBeforeButton, "Time Before"),
            (topScoresButton, "Top Scores")
        ]
        for (button, label) in buttonLabels { button?.accessibilityLabel = label }
        clipArtImageView?.isAccessibilityElement = false
        logoImageView?.isAccessibilityElement = false
        clipArtView?.isAccessibilityElement = false
        // (clockView is a ClockView; it already self-describes its time to VoiceOver.)
        if let cv = clipArtView {
            let clipArtFrame = CGRect(x: 0, y: 0, width: cv.frame.size.width, height: cv.frame.size.height)
            clipArtImageView?.frame = clipArtFrame
            logoImageView?.frame = clipArtFrame
            clipArtImageView?.contentMode = .scaleAspectFit
            logoImageView?.contentMode = .scaleAspectFit
            if let logo = logoImageView { cv.addSubview(logo) }
            clipArtTimer = Timer.scheduledTimer(timeInterval: 2, target: self, selector: #selector(changeClipArt), userInfo: nil, repeats: true)
        }
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        updatePlayerChip()
        refreshClock()
        clockTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(refreshClock), userInfo: nil, repeats: true)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        // At launch, show "Who's Playing?" unless there's exactly one player (then
        // just use them). 0 players → add the first; 2+ → choose.
        guard !MenuViewController.didPresentPickerThisLaunch else { return }
        MenuViewController.didPresentPickerThisLaunch = true
        if ProfileStore.shared.profiles.count != 1 || ProfileStore.shared.activeProfile == nil {
            presentProfilePicker(animated: false)
        }
    }

    // MARK: - Player profiles

    private func updatePlayerChip() {
        let p = ProfileStore.shared.activeProfile
        playerChip.setTitle(p.map { "\($0.avatar)  \($0.name)  ▾" } ?? kStrAppTitle, for: .normal)
        playerChip.setTitleColor(UIColor(red: 0.055, green: 0.478, blue: 0.996, alpha: 1), for: .normal)
        playerChip.titleLabel?.font = .boldSystemFont(ofSize: 18)
        playerChip.accessibilityLabel = p.map { "Player: \($0.name). Double tap to switch player." } ?? kStrAppTitle
        playerChip.sizeToFit()
    }

    @objc private func playerChipTapped() {
        presentProfilePicker(animated: true)
    }

    private func presentProfilePicker(animated: Bool) {
        guard presentedViewController == nil else { return }
        let nav = UINavigationController(rootViewController: ProfileSelectViewController())
        nav.modalPresentationStyle = .fullScreen
        present(nav, animated: animated)
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        clockTimer?.invalidate(); clockTimer = nil
    }

    deinit {
        clipArtTimer?.invalidate()
    }

    @objc func changeClipArt() {
        guard let cv = clipArtView, let civ = clipArtImageView else { return }
        let r = RandomInteger(range: Int(kClipArtFileRangeLow), to: Int(kClipArtFileRangeHigh))
        let name = String(format: kClipArtFileMask, r.randomInteger, kClipArtFileType)
        civ.image = UIImage(named: name)
        if let first = cv.subviews.first {
            cv.replaceSubview(first, withSubview: civ, transition: .push, direction: .fromLeft, duration: 0.10)
        }
    }

    @objc func refreshClock() {
        let cal = Calendar.autoupdatingCurrent
        let c = cal.dateComponents([.hour, .minute, .second], from: Date())
        clockView.hours = Float(c.hour ?? 0)
        clockView.minutes = Float(c.minute ?? 0)
        clockView.seconds = Float(c.second ?? 0)
        clockView.showSeconds = true; clockView.showClockAsAnalog = true
        clockView.showMinutesOffsetInHoursHand = true; clockView.showAMPM = false; clockView.showDayNight = false
        clockView.setNeedsDisplay()
    }

    @IBAction func tellTimeButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActTellTime
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction func setTimeButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActSetTime
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction func elapsedTimeButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActElapsedTime
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction func tellTimeAfterButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActTimeAfter
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction func tellTimeBeforeButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActTimeBefore
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction func mixedModeButtonPressed(_ sender: Any) {
        KidsTimeFunAppState.sharedState().activity = kActMixed
        navigationController?.pushViewController(activityVC, animated: true)
    }

    @IBAction @objc func topScoresButtonPressed(_ sender: Any) {
        navigationController?.pushViewController(topScoresActVC, animated: true)
    }

    @IBAction func setActivityType(_ sender: UISegmentedControl) {
        KidsTimeFunAppState.sharedState().activityType = Int32(sender.selectedSegmentIndex)
    }

    @IBAction @objc func settingsActivated() {
        guard navigationController?.topViewController == self else { return }
        navigationController?.pushViewController(settingsVC, animated: true)
    }

    @objc private func goHome() {}
}
